python/ray/dag/compiled_dag_node.py

```diff
@@ -907,11 +900,37 @@ def run(self):
         monitor.start()
         return monitor
 
+    def _execute_until(
+        self,
+        execution_index: int,
+    ) -> Any:
+        """Repeatedly execute this DAG until the given execution index.
+
+        Args:
+            execution_index: The execution index to execute until.
+
+        Returns:
+            The execution result corresponding to the given execution index.
+        """
+        while execution_index >= len(self._result_buffer):
+            if self._buffered_results > MAX_BUFFER_COUNT:
+                raise ValueError(
+                    "Too many buffered results: call ray.get() on previous "
+                    "CompiledDAGRefs to free them up from buffer."
+                )
+            self._result_buffer.append(self._dag_output_fetcher.begin_read())
+            self._buffered_results += 1
+
+        result = self._result_buffer[execution_index]
+        self._result_buffer[execution_index] = None
```

Comment:
This is to address out-of-order `get`. It's possible that result for execution_index 2 is obtained after execution_index 1. So essentially we need a value for each index to indicate if a value is buffered (not None) or already retrieved (None). And a list is more efficient than a dict.

I can add a TODO to limit the number of out-of-order executions, and we can cap the number of Nones.

Reply:
It should be possible to do this with a dict and a counter, right? I.e. if the index is not in the dict, and the counter is already past the index, then it is already retrieved. That way we can pop the results from the dict instead of accumulating a bunch of placeholder None's.

Let's do this in this PR; I don't think the change is large enough to merit a follow-up.

Reply:
SG. Done.